zkevm-circuits/src/anchor_tx_circuit/sign_verify.rs

```diff
@@ -0,0 +1,424 @@
+use crate::{
+    evm_circuit::util::{
+        constraint_builder::{BaseConstraintBuilder, ConstrainBuilderCommon},
+        split_u256_limb64,
+    },
+    table::{TxFieldTag, TxTable},
+    util::Challenges,
+};
+use eth_types::{geth_types::Transaction, Field, ToBigEndian, Word, U256};
+use gadgets::{
+    is_equal::IsEqualChip,
+    mul_add::{MulAddChip, MulAddConfig},
+    util::{split_u256, Expr},
+};
+use halo2_proofs::{
+    circuit::{Layouter, Region, Value},
+    plonk::{Advice, Column, ConstraintSystem, Error, Expression, Fixed, SecondPhase, Selector},
+    poly::Rotation,
+};
+use log::error;
+use once_cell::sync::Lazy;
+
+const ANCHOR_TX_ID: usize = 0;
+const MAX_DEGREE: usize = 10;
+const BYTE_POW_BASE: u64 = 1 << 8;
+
+// 0x79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798
+const GX1: Word = U256([
+    0x59F2815B16F81798,
+    0x029BFCDB2DCE28D9,
+    0x55A06295CE870B07,
+    0x79BE667EF9DCBBAC,
+]);
+static GX1_LO_HI: Lazy<(U256, U256)> = Lazy::new(|| split_u256(&GX1));
+
+// 0xc6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5
+const GX2: Word = U256([
+    0xabac09b95c709ee5,
+    0x5c778e4b8cef3ca7,
+    0x3045406e95c07cd8,
+    0xc6047f9441ed7d6d,
+]);
+static GX2_LO_HI: Lazy<(U256, U256)> = Lazy::new(|| split_u256(&GX2));
+
+// 0xfffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141
+const N: Word = U256([
+    0xbfd25e8cd0364141,
+    0xbaaedce6af48a03b,
+    0xfffffffffffffffe,
+    0xffffffffffffffff,
+]);
+static N_LO_HI: Lazy<(U256, U256)> = Lazy::new(|| split_u256(&N));
+
+// private key 0x92954368afd3caa1f3ce3ead0069c1af414054aefe1ef9aeacc1bf426222ce38
+// GX1 * PRIVATEKEY(mod N) = 0x4341adf5a780b4a87939938fd7a032f6e6664c7da553c121d3b4947429639122
+const GX1_MUL_PRIVATEKEY: Word = U256([
+    0xd3b4947429639122,
+    0xe6664c7da553c121,
+    0x7939938fd7a032f6,
+    0x4341adf5a780b4a8,
+]);
+static GX1_MUL_PRIVATEKEY_LO_HI: Lazy<(U256, U256)> = Lazy::new(|| split_u256(&GX1_MUL_PRIVATEKEY));
+static GX1_MUL_PRIVATEKEY_LIMB64: Lazy<[U256; 4]> =
+    Lazy::new(|| split_u256_limb64(&GX1_MUL_PRIVATEKEY));
+
+// GX2 * PRIVATEKEY(mod N) = 0x4a43b192ca74cab200d6c086df90fb729abca9e52d38b8fa0beb4eafe70956de
+const GX2_MUL_PRIVATEKEY: Word = U256([
+    0x0beb4eafe70956de,
+    0x9abca9e52d38b8fa,
+    0x00d6c086df90fb72,
+    0x4a43b192ca74cab2,
+]);
+static GX2_MUL_PRIVATEKEY_LO_HI: Lazy<(U256, U256)> = Lazy::new(|| split_u256(&GX2_MUL_PRIVATEKEY));
+
+// # How to check the signature
+// 1. IF r == GX1 OR r == GX2
+// 2. IF r == GX2 THEN MUST WHEN r == GX1 AND s == 0
+// 3. IF s == 0 THEN (GX1_MUL_PRIVATEKEY + msg_hash) == N
```

Comment:
Could the `(GX1_MUL_PRIVATEKEY + msg_hash) == N` be explained a little bit more?

Reply:
based on algrithm of [taiko-mono](https://github.com/taikoxyz/taiko-mono/blob/ad26803e5bcbcc76b812084b7bd08f45992e59dd/packages/protocol/contracts/libs/LibAnchorSignature.sol#L68)
```
s = (GX1 * GOLDEN_TOUCH_PRIVATEKEY + msg_hash) (mod N)
```
So, it equals to
```
s = (GX1 * GOLDEN_TOUCH_PRIVATEKEY (mod N) + msg_hash (mod N)) (mod N)
```
-  Our `GX1_MUL_PRIVATEKEY` is equal to `GX1 * GOLDEN_TOUCH_PRIVATEKEY (mod N)`
-  Our `msg_hash` has already been (mod N) in [zkevm-circuit](https://github.com/taikoxyz/zkevm-circuits/blob/839152c04ab3ddd1b8ce32632a407e5e7ef823a8/eth-types/src/geth_types.rs#L236)
```rust
let msg_hash = msg_hash.mod_floor(&*SECP256K1_Q);
```
Finally, we only need prove:
```
because: 0 < GX1_MUL_PRIVATEKEY + msg_hash < 2N
need prove: GX1_MUL_PRIVATEKEY + msg_hash  (mod N) == 0
so: GX1_MUL_PRIVATEKEY + msg_hash == N
```

Reply:
Would be great if you could add a bit more of this explanation inside the comments so it's directly available for everyone looking at the code.